syntax = "proto3";

package nitric.proto.storage.v2;

import "google/protobuf/duration.proto";

// protoc plugin options for code generation
option go_package = "github.com/nitrictech/nitric/proto/storage/v2;storagepb";
option java_package = "io.nitric.proto.storage.v2";
option java_multiple_files = true;
option java_outer_classname = "StorageService";
option php_namespace = "Nitric\\Proto\\Storage\\V2";
option csharp_namespace = "Nitric.Proto.Storage.v2";

// Services for storage and retrieval of blobs in the form of byte arrays, such as text and binary blobs.
service Storage {
  // Retrieve an item from a bucket
  rpc Read (StorageReadRequest) returns (StorageReadResponse);
  // Store an item to a bucket
  rpc Write (StorageWriteRequest) returns (StorageWriteResponse);
  // Delete an item from a bucket
  rpc Delete (StorageDeleteRequest) returns (StorageDeleteResponse);
  // Generate a pre-signed URL for direct operations on an item
  rpc PreSignUrl (StoragePreSignUrlRequest) returns (StoragePreSignUrlResponse);
  // List blobs currently in the bucket
  rpc ListBlobs (StorageListBlobsRequest) returns (StorageListBlobsResponse);
  // Determine is an object exists in a bucket
  rpc Exists (StorageExistsRequest) returns (StorageExistsResponse);
}

// Request to put (create/update) a storage item
message StorageWriteRequest {
  // Nitric name of the bucket to store in
  //  this will be automatically resolved to the provider specific bucket identifier.
  string bucket_name = 1;
  // Key to store the item under
  string key = 2;
  // bytes array to store
  bytes body = 3;
}

// Result of putting a storage item
message StorageWriteResponse {
}

// Request to retrieve a storage item
message StorageReadRequest {
  // Nitric name of the bucket to retrieve from
  //  this will be automatically resolved to the provider specific bucket identifier.
  string bucket_name = 1;
  // Key of item to retrieve
  string key = 2;
}

// Returned storage item
message StorageReadResponse {
  // The body bytes of the retrieved storage item
  bytes body = 1;
}

// Request to delete a storage item
message StorageDeleteRequest {
  // Name of the bucket to delete from
  string bucket_name = 1;
  // Key of item to delete
  string key = 2;
}

// Result of deleting a storage item
message StorageDeleteResponse {
}

// Request to generate a pre-signed URL for a blob to perform a specific operation, such as read or write.
message StoragePreSignUrlRequest {
  // Nitric name of the bucket to retrieve from
  //  this will be automatically resolved to the provider specific bucket identifier.
  string bucket_name = 1;
  // Key of item to generate the signed URL for.
  // The URL and the token it contains will only be valid for operations on this resource specifically.
  string key = 2;
  // Operation
  enum Operation {
    READ = 0;
    WRITE = 1;
  }
  Operation operation = 3;
  // Expiry defined as as protobuf duration
  google.protobuf.Duration expiry = 4;
}

message StoragePreSignUrlResponse {
  // The pre-signed url, restricted to the operation, resource and expiry time specified in the request.
  string url = 1;
}

message StorageListBlobsRequest {
  string bucket_name = 1;

  string prefix = 2;
}

message Blob {
  string key = 1;
}

message StorageListBlobsResponse {
  // keys of the blobs in the bucket
  repeated Blob blobs = 1;
}

message StorageExistsRequest {
  string bucket_name = 1;
  // Key of item to retrieve
  string key = 2;
}

message StorageExistsResponse {
  bool exists = 1;
}
